File: backend/internal/models/product.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// Category represents a product category (e.g. Floral, Woody, Oriental).
type Category struct {
	ID          uuid.UUID `json:"id"`
	Name        string    `json:"name"`
	Description *string   `json:"description,omitempty"`
	SortOrder   int       `json:"sort_order"`
	CreatedAt   time.Time `json:"created_at"`
}

// ProductVariant represents one size/price/stock option of a product.
// e.g. Rose Elégante comes in 30ml, 50ml, 100ml — each is a variant.
type ProductVariant struct {
	ID        uuid.UUID `json:"id"`
	ProductID uuid.UUID `json:"product_id"`
	Size      string    `json:"size"`
	Price     int64     `json:"price"` // stored in IDR (e.g. 185000)
	Stock     int       `json:"stock"`
	SKU       string    `json:"sku"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// Product represents a perfume product with its variants.
type Product struct {
	ID          uuid.UUID        `json:"id"`
	Name        string           `json:"name"`
	Brand       string           `json:"brand"`
	CategoryID  uuid.UUID        `json:"category_id"`
	Category    *Category        `json:"category,omitempty"`
	Description *string          `json:"description,omitempty"`
	ImageURL    *string          `json:"image_url,omitempty"`
	IsActive    bool             `json:"is_active"`
	IsFeatured  bool             `json:"is_featured"`
	IsFavourite bool             `json:"is_favourite"`
	Variants    []ProductVariant `json:"variants"`
	CreatedAt   time.Time        `json:"created_at"`
	UpdatedAt   time.Time        `json:"updated_at"`
}

// TotalStock returns the sum of stock across all variants.
// Used by the cashier to quickly know if a product is available.
func (p *Product) TotalStock() int {
	total := 0
	for _, v := range p.Variants {
		total += v.Stock
	}
	return total
}

// IsOutOfStock returns true if no variants have stock.
func (p *Product) IsOutOfStock() bool {
	return p.TotalStock() == 0
}

// ── Request bodies (what the API accepts) ──────────────────────

// CreateProductRequest is the JSON body for POST /products.
type CreateProductRequest struct {
	Name        string                 `json:"name"`
	Brand       string                 `json:"brand"`
	CategoryID  string                 `json:"category_id"`
	Description string                 `json:"description"`
	ImageURL    string                 `json:"image_url"`
	IsActive    bool                   `json:"is_active"`
	IsFeatured  bool                   `json:"is_featured"`
	Variants    []CreateVariantRequest `json:"variants"`
}

// CreateVariantRequest is the variant data inside CreateProductRequest.
type CreateVariantRequest struct {
	Size  string `json:"size"`
	Price int64  `json:"price"`
	Stock int    `json:"stock"`
	SKU   string `json:"sku"`
}

// UpdateProductRequest is the JSON body for PATCH /products/{id}.
// Pointer fields mean "only update if provided in the request".
type UpdateProductRequest struct {
	Name        *string `json:"name"`
	Brand       *string `json:"brand"`
	CategoryID  *string `json:"category_id"`
	Description *string `json:"description"`
	ImageURL    *string `json:"image_url"`
	IsActive    *bool   `json:"is_active"`
	IsFeatured  *bool   `json:"is_featured"`
	IsFavourite *bool   `json:"is_favourite"`
}

// CreateCategoryRequest is the JSON body for POST /categories.
type CreateCategoryRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	SortOrder   int    `json:"sort_order"`
}

File: flutter_app/lib/features/products/models/product.dart
// These models mirror the Go backend's JSON responses.
// The field names here must match exactly what the API returns.
import 'package:flutter/foundation.dart';

@immutable
class Category {
  final String id;
  final String name;
  final String? description;
  final int sortOrder;

  const Category({
    required this.id,
    required this.name,
    this.description,
    required this.sortOrder,
  });

  factory Category.fromJson(Map<String, dynamic> json) => Category(
        id: json['id'] as String,
        name: json['name'] as String,
        description: json['description'] as String?,
        sortOrder: json['sort_order'] as int? ?? 0,
      );
}

@immutable
class ProductVariant {
  final String id;
  final String productId;
  final String size;
  final int price;   // IDR, e.g. 185000
  final int stock;
  final String sku;

  const ProductVariant({
    required this.id,
    required this.productId,
    required this.size,
    required this.price,
    required this.stock,
    required this.sku,
  });

  factory ProductVariant.fromJson(Map<String, dynamic> json) => ProductVariant(
        id: json['id'] as String,
        productId: json['product_id'] as String,
        size: json['size'] as String,
        price: (json['price'] as num).toInt(),
        stock: json['stock'] as int,
        sku: json['sku'] as String? ?? '',
      );

  bool get isOutOfStock => stock == 0;
  bool get isLowStock => stock > 0 && stock <= 5;
}

@immutable
class Product {
  final String id;
  final String name;
  final String brand;
  final String categoryId;
  final Category? category;
  final String? description;
  final String? imageUrl;
  final bool isActive;
  final bool isFeatured;
  final bool isFavourite;
  final List<ProductVariant> variants;

  const Product({
    required this.id,
    required this.name,
    required this.brand,
    required this.categoryId,
    this.category,
    this.description,
    this.imageUrl,
    required this.isActive,
    required this.isFeatured,
    required this.isFavourite,
    required this.variants,
  });

  factory Product.fromJson(Map<String, dynamic> json) => Product(
        id: json['id'] as String,
        name: json['name'] as String,
        brand: json['brand'] as String,
        categoryId: json['category_id'] as String,
        category: json['category'] != null
            ? Category.fromJson(json['category'] as Map<String, dynamic>)
            : null,
        description: json['description'] as String?,
        imageUrl: json['image_url'] as String?,
        isActive: json['is_active'] as bool? ?? true,
        isFeatured: json['is_featured'] as bool? ?? false,
        isFavourite: json['is_favourite'] as bool? ?? false,
        variants: (json['variants'] as List<dynamic>? ?? [])
            .map((v) => ProductVariant.fromJson(v as Map<String, dynamic>))
            .toList(),
      );

  int get totalStock => variants.fold(0, (sum, v) => sum + v.stock);
  bool get isOutOfStock => totalStock == 0;
}